In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
pd.set_option('display.max_columns', None)

In [3]:
# bld = pd.read_csv('./data/bank_loan_data.csv')
# print(bld.shape)
# bld.head()

In [4]:
# lac = pd.read_csv('./data/loan_approval_class.csv')
# print(lac.shape)
# lac.head()

# EQUAL TO BLD

In [5]:
# lad = pd.read_csv('./data/loan_approval_dataset.csv')
# print(lad.shape)
# lad.head()

In [6]:
la = pd.read_csv('./data/loan_approval.csv')
print(la.shape)
la.head()

(1000, 20)


Applicant_ID  Applicant_Income  Coapplicant_Income Employment_Status   Age  \
0           1.0           17795.0              1387.0          Salaried  51.0   
1           2.0            2860.0              2679.0          Salaried  46.0   
2           3.0            7390.0              2106.0          Salaried  25.0   
3           4.0           13964.0              8173.0          Salaried  40.0   
4           5.0           13284.0              4223.0     Self-employed  31.0   

  Marital_Status  Dependents  Credit_Score  Existing_Loans  DTI_Ratio  \
0        Married         0.0         637.0             4.0       0.53   
1        Married         3.0         621.0             2.0       0.30   
2         Single         2.0         674.0             4.0       0.20   
3        Married         2.0         579.0             3.0       0.31   
4         Single         2.0         721.0             1.0       0.29   

   Savings  Collateral_Value  Loan_Amount  Loan_Term Loan_Purpose  \
0  19403.0           45638.0      16619.0       84.0     Personal   
1   2580.0           49272.0      38687.0        NaN          Car   
2  13844.0            6908.0      27943.0       72.0          NaN   
3   9553.0           10844.0      27819.0       60.0     Business   
4   9386.0           37629.0      12741.0       72.0          Car   

  Property_Area Education_Level  Gender Employer_Category Loan_Approved  
0         Urban    Not Graduate  Female           Private            No  
1     Semiurban        Graduate     NaN           Private            No  
2         Urban             NaN  Female        Government           Yes  
3         Rural        Graduate  Female        Government            No  
4           NaN        Graduate    Male           Private           Yes

In [7]:
ld = pd.read_csv('./data/loan_data.csv')
print(ld.shape)
ld.head()

(614, 12)


Gender Married Dependents     Education Self_Employed  Applicant_Income  \
0   Male      No          0      Graduate            No            584900   
1   Male     Yes          1      Graduate            No            458300   
2   Male     Yes          0      Graduate           Yes            300000   
3   Male     Yes          0  Not Graduate            No            258300   
4   Male      No          0      Graduate            No            600000   

   Coapplicant_Income  Loan_Amount   Term  Credit_History   Area Status  
0                 0.0     15000000  360.0             1.0  Urban      Y  
1            150800.0     12800000  360.0             1.0  Rural      N  
2                 0.0      6600000  360.0             1.0  Urban      Y  
3            235800.0     12000000  360.0             1.0  Urban      Y  
4                 0.0     14100000  360.0             1.0  Urban      Y

## Common Columns

- Gender
- Marital Status
- Area/Property Area
- Loan Amount
- Credit History (requires feature engineering)
- Loan Term
- Applicant Income
- Coapplicant Income
- Loan Status
- Education Level

Let's modify the datasets so that they match each other, than one dataset can be used for training and preliminary testing, and the other purely for testing. But first, both sets need to be cleaned of missing data and obscure outliers.  

## Data Cleaning

### loan_approval.csv

In [8]:
df = pd.read_csv('./data/loan_approval.csv')

In [9]:
df.head()

Applicant_ID  Applicant_Income  Coapplicant_Income Employment_Status   Age  \
0           1.0           17795.0              1387.0          Salaried  51.0   
1           2.0            2860.0              2679.0          Salaried  46.0   
2           3.0            7390.0              2106.0          Salaried  25.0   
3           4.0           13964.0              8173.0          Salaried  40.0   
4           5.0           13284.0              4223.0     Self-employed  31.0   

  Marital_Status  Dependents  Credit_Score  Existing_Loans  DTI_Ratio  \
0        Married         0.0         637.0             4.0       0.53   
1        Married         3.0         621.0             2.0       0.30   
2         Single         2.0         674.0             4.0       0.20   
3        Married         2.0         579.0             3.0       0.31   
4         Single         2.0         721.0             1.0       0.29   

   Savings  Collateral_Value  Loan_Amount  Loan_Term Loan_Purpose  \
0  19403.0           45638.0      16619.0       84.0     Personal   
1   2580.0           49272.0      38687.0        NaN          Car   
2  13844.0            6908.0      27943.0       72.0          NaN   
3   9553.0           10844.0      27819.0       60.0     Business   
4   9386.0           37629.0      12741.0       72.0          Car   

  Property_Area Education_Level  Gender Employer_Category Loan_Approved  
0         Urban    Not Graduate  Female           Private            No  
1     Semiurban        Graduate     NaN           Private            No  
2         Urban             NaN  Female        Government           Yes  
3         Rural        Graduate  Female        Government            No  
4           NaN        Graduate    Male           Private           Yes

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Applicant_ID        950 non-null    float64
 1   Applicant_Income    950 non-null    float64
 2   Coapplicant_Income  950 non-null    float64
 3   Employment_Status   950 non-null    object 
 4   Age                 950 non-null    float64
 5   Marital_Status      950 non-null    object 
 6   Dependents          950 non-null    float64
 7   Credit_Score        950 non-null    float64
 8   Existing_Loans      950 non-null    float64
 9   DTI_Ratio           950 non-null    float64
 10  Savings             950 non-null    float64
 11  Collateral_Value    950 non-null    float64
 12  Loan_Amount         950 non-null    float64
 13  Loan_Term           950 non-null    float64
 14  Loan_Purpose        950 non-null    object 
 15  Property_Area       950 non-null    object 
 16  Educati

In [11]:
df.isna().sum()

Applicant_ID          50
Applicant_Income      50
Coapplicant_Income    50
Employment_Status     50
Age                   50
Marital_Status        50
Dependents            50
Credit_Score          50
Existing_Loans        50
DTI_Ratio             50
Savings               50
Collateral_Value      50
Loan_Amount           50
Loan_Term             50
Loan_Purpose          50
Property_Area         50
Education_Level       50
Gender                50
Employer_Category     50
Loan_Approved         50
dtype: int64

In [12]:
rows_missing_any = df[df.isna().sum(axis=1) >= 1]

In [13]:
rows_missing_any.shape

(649, 20)

There are 649 rows that are missing at least one value. 

Let's see what columns we can definately drop (are not needed for loan approval evaluation). 

We also listed above which columns are the same in both of our useful datasets, so we can drop the columns that do not match

In [14]:
df.columns

Index(['Applicant_ID', 'Applicant_Income', 'Coapplicant_Income',
       'Employment_Status', 'Age', 'Marital_Status', 'Dependents',
       'Credit_Score', 'Existing_Loans', 'DTI_Ratio', 'Savings',
       'Collateral_Value', 'Loan_Amount', 'Loan_Term', 'Loan_Purpose',
       'Property_Area', 'Education_Level', 'Gender', 'Employer_Category',
       'Loan_Approved'],
      dtype='object')

In [15]:
df = df.drop(columns=['Applicant_ID', 'Employment_Status', 'Age', 'Dependents', 'Existing_Loans', 'DTI_Ratio', 'Savings', 'Collateral_Value', 'Loan_Purpose' ,'Employer_Category'])

In [16]:
rows_missing_any = df[df.isna().sum(axis=1) >= 1]

In [17]:
rows_missing_any.shape

(392, 10)

In [18]:
df.columns

Index(['Applicant_Income', 'Coapplicant_Income', 'Marital_Status',
       'Credit_Score', 'Loan_Amount', 'Loan_Term', 'Property_Area',
       'Education_Level', 'Gender', 'Loan_Approved'],
      dtype='object')

In [19]:
df.isna().sum()

Applicant_Income      50
Coapplicant_Income    50
Marital_Status        50
Credit_Score          50
Loan_Amount           50
Loan_Term             50
Property_Area         50
Education_Level       50
Gender                50
Loan_Approved         50
dtype: int64

We cannot impute Loan_Approved, so we will drop those NA values at the end. For the rest of the data, we will look at each one to determine how best to impute the missing values. 

This will be done in another notebook. 

In [20]:
df.to_csv('./data_processing/loan_approval.csv', index=False)

### loan_data.csv

In [21]:
df = pd.read_csv('./data/loan_data.csv')

In [22]:
df.head()

Gender Married Dependents     Education Self_Employed  Applicant_Income  \
0   Male      No          0      Graduate            No            584900   
1   Male     Yes          1      Graduate            No            458300   
2   Male     Yes          0      Graduate           Yes            300000   
3   Male     Yes          0  Not Graduate            No            258300   
4   Male      No          0      Graduate            No            600000   

   Coapplicant_Income  Loan_Amount   Term  Credit_History   Area Status  
0                 0.0     15000000  360.0             1.0  Urban      Y  
1            150800.0     12800000  360.0             1.0  Rural      N  
2                 0.0      6600000  360.0             1.0  Urban      Y  
3            235800.0     12000000  360.0             1.0  Urban      Y  
4                 0.0     14100000  360.0             1.0  Urban      Y

In [23]:
df.isna().sum()

Gender                13
Married                3
Dependents            15
Education              0
Self_Employed         32
Applicant_Income       0
Coapplicant_Income     0
Loan_Amount            0
Term                  14
Credit_History        50
Area                   0
Status                 0
dtype: int64

In [24]:
df.shape

(614, 12)

Again, let's drop the columns we don't need or can't use. 

In [25]:
df = df.drop(columns=['Dependents', 'Self_Employed'])

In [26]:
df.isna().sum()

Gender                13
Married                3
Education              0
Applicant_Income       0
Coapplicant_Income     0
Loan_Amount            0
Term                  14
Credit_History        50
Area                   0
Status                 0
dtype: int64

Again, we will do the cleaning in another notebook. 

In [27]:
df.to_csv('./data_processing/loan_data.csv', index=False)